### PR TITLE
Persistence Test: Introduce AllRowsEquivalentToJson and ActiveRowsEquivalentToJson

### PR DESCRIPTION
**Pull request by kumuwu**
## Summary

Adds two new types of TestAssertion: AllRowsEquivalentToJson and ActiveRowsEquivalentToJson (and the corresponding fail statuses), to be used by persistence tests. 
Depends on https://github.com/finos/legend-engine/pull/1242 

## How did you test this change?

- [x] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

Screenshot 1: test passes case

Success

Screenshot 2: test fails case

Fail

### CODE REVIEW

File: .changeset/wicked-terms-push.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+'@finos/legend-extension-dsl-persistence': patch
+'@finos/legend-application-studio': patch
+'@finos/legend-graph': patch
+---
```
Comment (gayathrir11) [inline .changeset/wicked-terms-push.md:5]: Can you split these changesets please?
Have on changeset for `dsl-persistence` with description 
`Add new assertions types for persistence tests` 
and a different changeset for the other two packages with description `Add extension mechanism for TestAssertion`

File: packages/legend-extension-dsl-persistence/src/graphManager/protocol/pure/DSL_Persistence_PureProtocolProcessorPlugin.ts [hunk 2]
```diff
@@ -401,4 +528,130 @@ export class DSL_Persistence_PureProtocolProcessorPlugin
...
+        plugins: PureProtocolProcessorPlugin[],
+      ): AssertionStatus | undefined => {
+        if (element instanceof V1_AllRowsEquivalentToJsonAssertFail) {
+          let assertion = atomicTest.assertions.find(
+            (a) => a.id === element.id,
+          );
+          const extraAssertionBuilder = plugins.flatMap(
+            (plugin) =>
+              (
+                plugin as Testable_PureProtocolProcessorPlugin_Extension
```
Comment (gayathrir11) [inline packages/legend-extension-dsl-persistence/src/graphManager/protocol/pure/DSL_Persistence_PureProtocolProcessorPlugin.ts:546]: why do you call these plugins here? Is it to build the metamodel for `AllRowsEquivalentToJsonAssertion`?

File: packages/legend-extension-dsl-persistence/src/graphManager/protocol/pure/DSL_Persistence_PureProtocolProcessorPlugin.ts [hunk 3]
```diff
@@ -401,4 +528,130 @@ export class DSL_Persistence_PureProtocolProcessorPlugin
...
+            `Can't build AllRowsEquivalentToJsonAssertFail: no compatible builder available from plugins`,
+            element,
+          );
+        } else if (element instanceof V1_ActiveRowsEquivalentToJsonAssertFail) {
+          let assertion = atomicTest.assertions.find(
+            (a) => a.id === element.id,
+          );
+          const extraAssertionBuilder = plugins.flatMap(
+            (plugin) =>
+              (
```
Comment (gayathrir11) [inline packages/legend-extension-dsl-persistence/src/graphManager/protocol/pure/DSL_Persistence_PureProtocolProcessorPlugin.ts:574]: same as above

File: packages/legend-graph/src/graphManager/Testable_PureGraphManagerPlugin_Extension.ts [hunk 4]
```diff
@@ -51,4 +57,9 @@ export interface Testable_PureGraphManagerPlugin_Extension
    * Get the list of atomic test observers.
    */
   getExtraAtomicTestObservers?(): AtomicTestObserver[];
+
+  /**
+   * Get the list of atomic test observers.
+   */
```
Comment (gayathrir11) [inline packages/legend-graph/src/graphManager/Testable_PureGraphManagerPlugin_Extension.ts:63]: Can you please change the comment here according to what you added

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_TestBuilderHelper.ts [hunk 5]
```diff
@@ -92,6 +92,18 @@ export const V1_buildTestAssertion = (
   } else if (value instanceof V1_EqualToTDS) {
     return buildEqualToTDS(value, parentTest, context);
   }
+  const extraTestAssertionBuilder = context.extensions.plugins.flatMap(
+    (plugin) =>
```
Comment (gayathrir11) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_TestBuilderHelper.ts:96]: --> `extraAssertionBuilders`

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_TestResultBuilderHelper.ts [hunk 6]
```diff
@@ -178,6 +178,18 @@ const buildAssertionStatus = (
   } else if (value instanceof V1_AssertPass) {
     return buildAssertPass(value, atomicTest, plugins);
   }
+  const extraAssertionStatusBuilder = plugins.flatMap(
+    (plugin) =>
```
Comment (gayathrir11) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_TestResultBuilderHelper.ts:182]: --> `extraAssertionStatusBuilders`

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureProtocol/serializationHelpers/V1_TestSerializationHelper.ts [hunk 7]
```diff
@@ -145,10 +160,23 @@ const V1_deserializeAssertionStatus = (
...
+          ).V1_getExtraAssertionStatusProtocolDeserializers?.() ?? [],
+      );
+      for (const deserializer of extraAssertionStatusProtocolDeserializers) {
+        const assertionStatusProtocol = deserializer(json, plugins);
+        if (assertionStatusProtocol) {
+          return assertionStatusProtocol;
+        }
+      }
       throw new UnsupportedOperationError(
         `Can't deserialize assertion status of type '${json._type}'`,
```
Comment (gayathrir11) [inline packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureProtocol/serializationHelpers/V1_TestSerializationHelper.ts:177]: Would you mind updating the message to similar format in places wherever you added the extension mechanism in the code.
`Can't deserialize assertion status of type '${json._type}': no compatible serializer available from the plugins`

**Comment by gayathrir11** [comment]
Can you some grammar tests in `legend-manual-tests` and some unit tests in `DSL_Persistence_Roundtrip.test.ts`

File: .changeset/wicked-terms-push.md (hunk 1 above)
Comment (kumuwu) [reply .changeset/wicked-terms-push.md:3]: Sure, changed!

File: packages/legend-extension-dsl-persistence/src/graphManager/protocol/pure/DSL_Persistence_PureProtocolProcessorPlugin.ts (hunk 2 above)
Comment (kumuwu) [reply packages/legend-extension-dsl-persistence/src/graphManager/protocol/pure/DSL_Persistence_PureProtocolProcessorPlugin.ts:546]: Here it is building extra testable assertions, not assertion status. The assertion that is retrieved and built from the plugins is then used in the building of the assertion status (line 557 - 563).   
I got this pattern from all the other assertion statuses - seems like this extra step is needed for retrieving the correct assertion to be used. Please refer to `buildEqualToJsonAssertFail` (or any other builder for assertion statuses) in V1_TestResultBuilderHelper.ts`

File: packages/legend-extension-dsl-persistence/src/graphManager/protocol/pure/DSL_Persistence_PureProtocolProcessorPlugin.ts (hunk 3 above)
Comment (kumuwu) [reply packages/legend-extension-dsl-persistence/src/graphManager/protocol/pure/DSL_Persistence_PureProtocolProcessorPlugin.ts:574]: Same comment as above too

File: packages/legend-graph/src/graphManager/Testable_PureGraphManagerPlugin_Extension.ts (hunk 4 above)
Comment (kumuwu) [reply packages/legend-graph/src/graphManager/Testable_PureGraphManagerPlugin_Extension.ts:63]: Sure, changed!

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_TestBuilderHelper.ts (hunk 5 above)
Comment (kumuwu) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_TestBuilderHelper.ts:96]: Sure, changed!

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_TestResultBuilderHelper.ts (hunk 6 above)
Comment (kumuwu) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureGraph/to/helpers/V1_TestResultBuilderHelper.ts:182]: Sure, changed!

File: packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureProtocol/serializationHelpers/V1_TestSerializationHelper.ts (hunk 7 above)
Comment (kumuwu) [reply packages/legend-graph/src/graphManager/protocol/pure/v1/transformation/pureProtocol/serializationHelpers/V1_TestSerializationHelper.ts]: Sure, changed!

**Comment by kumuwu** [comment]
> Can you add some grammar tests in `legend-manual-tests` and some unit tests in `DSL_Persistence_Roundtrip.test.ts`

Added tests in `legend-manual-tests`. For `DSL_Persistence_Roundtrip.test.ts`, I have already included tests, please let me know if they are okay :)

**Review by gayathrir11 — APPROVED** [review]
Looks good to me!

**Pull request closed without merging**